### PR TITLE
[DO NOT MERGE] Suecreek 20190104

### PR DESCRIPTION
**Pull request by lyakh**
WIP, not for merging, I'm still working to address comments and to test on top of #740, but the UART is working with its interrupt routed via DW INTC with this version.

### CODE REVIEW

File: src/drivers/intel/cavs/spi.c [hunk 1]
```diff
@@ -0,0 +1,461 @@
+/*
```
[the remark this answers is not in the document]
Comment (lyakh) [reply src/drivers/intel/cavs/spi.c]: You want all DW drivers in drivers/dw? Should I also move dw-dma there then?

File: src/drivers/intel/cavs/spi.c [hunk 2]
```diff
@@ -0,0 +1,461 @@
...
+
+#if 1 // TODO
+static void spi_dma_complete(void *data,
+				uint32_t type, struct dma_sg_elem *next)
+{
+	struct sspi *spi = (struct sspi *)data;
+	struct sof_ipc_hdr *hdr;
+	struct spi_dma_config *spi_config;
+
+	dcache_invalidate_region(spi->rx_buffer, SPI_BUFFER_SIZE);
```
[the remark this answers is not in the document]
Comment (lyakh) [reply src/drivers/intel/cavs/spi.c]: It is written back in spi_push(), why should it be done in completion?

File: src/drivers/intel/cavs/spi.c [hunk 3]
```diff
@@ -0,0 +1,461 @@
...
+}
+#endif
+
+static int spi_slave_probe(struct sspi *spi)
+{
+	spi->dma = dma_get(DMA_ID_DMAC0);
+	spi->chan[SPI_DIR_RX]  = dma_channel_get(spi->dma);
+	spi->chan[SPI_DIR_TX] = dma_channel_get(spi->dma);
+
+	/* configure the spi clock */
```
[the remark this answers is not in the document]
Comment (lyakh) [reply src/drivers/intel/cavs/spi.c]: If I knew that I'd gladly put it there ;-)

**Comment by lgirdwood** [comment]
@lyakh I'll close this since you are merging in smaller PRs.

**Pull request closed without merging**